web/analytics/ai_prediction.py

```
@@ -0,0 +1,251 @@
+            # Determine risk level based on score
+            if risk_score >= 70:
+                return "high", risk_factors
+            elif risk_score >= 40:
+                return "medium", risk_factors
+            else:
+                return "low", risk_factors
+                
+        except (Enrollment.DoesNotExist, CourseProgress.DoesNotExist):
+            return "high", ["No enrollment or progress data"]
+    
+    def generate_personalized_recommendations(self, student, courses=None):
```

Ruff (0.8.2)

Missing return type annotation for public function `generate_personalized_recommendations`

(ANN201)